api/e2e/test/router_test.go

```diff
@@ -111,3 +118,196 @@ var _ = DeployedRouterContext("testdata/create_router_with_traffic_rules.json.tm
+						conn, _ := grpc.Dial(routerCtx.Endpoint,
+							grpc.WithTransportCredentials(insecure.NewCredentials()))
+						defer conn.Close()
+
+						client := upiv1.NewUniversalPredictionServiceClient(conn)
+						headers := metadata.New(map[string]string{"region": "region-a"})
+						upiRequest := &upiv1.PredictValuesRequest{
+							PredictionTable: &upiv1.Table{
+								Name: "Test",
+								Columns: []*upiv1.Column{
+									{
+										Name: "col1",
+										Type: upiv1.Type_TYPE_DOUBLE,
+									},
+								},
+								Rows: []*upiv1.Row{
+									{
+										RowId: "1",
+										Values: []*upiv1.Value{
+											{},
+										},
+									},
+								},
+							},
+							PredictionContext: []*upiv1.Variable{
+								{
+									Name:        "client_id",
+									Type:        upiv1.Type_TYPE_STRING,
+									StringValue: "4",
+								},
+							},
+						}
+
+						resp, err := client.PredictValues(metadata.NewOutgoingContext(context.Background(), headers),
+							upiRequest)
+
+						Expect(err).To(BeNil())
+						Expect(resp.Metadata.Models[0].Name, err).To(Equal("treatment-a"))
+						Expect(resp.PredictionResultTable).To(matcher.ProtoEqual(upiRequest.PredictionTable))
+					})
+				})
+			})
+		})
+	})
+
+var _ = DeployedRouterContext("testdata/create_router_upi_with_traffic_rules.json.tmpl", routerConfig.UPI,
+	func(routerCtx *RouterContext) {
+		Describe("Calling UPI router endpoint to fetch predictions", func() {
+			Context("Turing Router API", func() {
+				When("send UPI PredictValues that satisfy traffic rule rule-1", func() {
+					It("responds successfully using treatment-a endpoint", func() {
+						conn, _ := grpc.Dial(routerCtx.Endpoint,
+							grpc.WithTransportCredentials(insecure.NewCredentials()))
+						defer conn.Close()
+
+						client := upiv1.NewUniversalPredictionServiceClient(conn)
+						headers := metadata.New(map[string]string{})
+						upiRequest := &upiv1.PredictValuesRequest{
+							PredictionTable: &upiv1.Table{
+								Name: "Test",
+								Columns: []*upiv1.Column{
+									{
+										Name: "col1",
+										Type: upiv1.Type_TYPE_DOUBLE,
+									},
+								},
+								Rows: []*upiv1.Row{
+									{
+										RowId: "1",
+										Values: []*upiv1.Value{
+											{},
+										},
+									},
+								},
+							},
+							PredictionContext: []*upiv1.Variable{
+								{
+									Name:        "client_id",
+									Type:        upiv1.Type_TYPE_STRING,
+									StringValue: "1",
+								},
+							},
+						}
+
+						resp, err := client.PredictValues(metadata.NewOutgoingContext(context.Background(), headers),
+							upiRequest)
+
+						Expect(err).To(BeNil())
+						Expect(resp.Metadata.Models[0].Name, err).To(Equal("treatment-a"))
+						Expect(resp.PredictionResultTable).To(matcher.ProtoEqual(upiRequest.PredictionTable))
+					})
+				})
+				When("send UPI PredictValues that that doesn't satisfy any traffic rule", func() {
```

Comment:
nit: `When("Sent UPI PredictValues that doesn't satisfy any traffic rule", ...`

Reply:
updated